allennlp/commands/__init__.py

```diff
@@ -51,37 +50,33 @@ def add_argument(self, *args, **kwargs):
 
 
 def create_parser(
-    prog: str = None, subcommand_overrides: Dict[str, Subcommand] = None
+    prog: str = None, subcommand_overrides: Iterable[Subcommand] = None
 ) -> argparse.ArgumentParser:
     """
     Creates the argument parser for the main program.
     """
-    if subcommand_overrides is None:
-        subcommand_overrides = {}
+    subcommand_overrides = subcommand_overrides or []
 
-    parser = ArgumentParserWithDefaults(description="Run AllenNLP", usage="%(prog)s", prog=prog)
-    parser.add_argument("--version", action="version", version="%(prog)s " + __version__)
+    parser = ArgumentParserWithDefaults(description="Run AllenNLP", prog=prog)
+    parser.add_argument("--version", action="version", version=f"%(prog)s {__version__}")
 
     subparsers = parser.add_subparsers(title="Commands", metavar="")
 
-    subcommands = {
-        # Default commands
-        "train": Train(),
-        "evaluate": Evaluate(),
-        "predict": Predict(),
-        "elmo": Elmo(),
-        "fine-tune": FineTune(),
-        "dry-run": DryRun(),
-        "make-vocab": DryRun(),  # deprecated, but keeping for backward compatibility.
-        "test-install": TestInstall(),
-        "find-lr": FindLearningRate(),
-        "print-results": PrintResults(),
-        # Superseded by overrides
-        **subcommand_overrides,
-    }
-
-    for name, subcommand in subcommands.items():
-        subparser = subcommand.add_subparser(name, subparsers)
+    subcommand_overrides_names = {subcommand.name for subcommand in subcommand_overrides}
+
+    subcommands = set(subcommand_overrides)
+
+    # The class of each instance in the overrides iterable is gonna be in the subclasses
+    # iterable, by definition. So we only add those `Subcommand` subclasses that don't have an
+    # instance in the overrides iterable.
+    for subcommand_class in Subcommand.__subclasses__():
+        subcommand = subcommand_class()
+        if subcommand.name not in subcommand_overrides_names:
+            subcommands.add(subcommand)
```

Comment:
Note that the overridden subcommands have preferences over the rest. However, there's no preference between the plugin's ones and the ones declared in allennlp. Is this okay?